### PR TITLE
Fixed #28184 -- Allowed passing a callable storage to FileField.

### PR DESCRIPTION
**Pull request by miigotu**
Adds the ability to pass a callable as the storage param of FileField, and testing for this new ability

See [ticket 28184](https://code.djangoproject.com/ticket/28184#ticket) for details.

Signed-off-by: miigotu <miigotu@gmail.com>

### CODE REVIEW

File: django/db/models/fields/files.py [hunk 1]
```diff
@@ -225,6 +225,10 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self._primary_key_set_explicitly = 'primary_key' in kwargs
 
         self.storage = storage or default_storage
+
+        if callable(self.storage):
+            self.storage = self.storage()
```
Comment (theodesp) [inline django/db/models/fields/files.py]: What happens if you don't return a valid storage type?

File: tests/file_storage/tests.py [hunk 2]
```diff
@@ -801,6 +804,10 @@ def test_stringio(self):
         with temp_storage.open('tests/stringio') as f:
             self.assertEqual(f.read(), b'content')
 
+    def test_callable_storage_param(self):
+        obj = Storage.objects.create()
```
Comment (theodesp) [inline tests/file_storage/tests.py]: The happy path is ok. What if you pass anything but a Storage object?

File: django/db/models/fields/files.py (hunk 1 above)
Comment (miigotu) [inline django/db/models/fields/files.py]: I assume it would blow up, and they would know right away that they did something wrong. Just as in other places of the code that accept callables.

File: tests/file_storage/tests.py (hunk 2 above)
Comment (miigotu) [inline tests/file_storage/tests.py]: Same answer

File: django/db/models/fields/files.py (hunk 1 above)
Comment (miigotu) [inline django/db/models/fields/files.py]: I could add an issubclass, if needed, but the added overhead may not be necessary.

**Comment by miigotu** [comment]
Yes, I'll fix the build first and then see about the other checks. About the type checking though, there is also no type checking in other places callables are allowed so I just did it the same way. I'll test what exactly DOES happen and see for sure.

File: tests/file_storage/tests.py [hunk 3]
```diff
@@ -807,6 +810,9 @@ def test_callable_storage_param(self):
         obj = Storage.objects.create()
         self.assertIs(obj.callable_storage.storage, temp_valid_name_storage)
 
+    def test_callable_base_class_error_raises(self):
+        self.assertRaises(ImproperlyConfigured, FileField(storage=lambda x:x))
```
Comment (carltongibson) [inline tests/file_storage/tests.py]: * Please use `assertRaisesMessage` here. 
* `flake8` will require whitespace after the `:` in the lambda.

File: django/db/models/fields/files.py [hunk 4]
```diff
@@ -226,7 +227,10 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self.storage = storage or default_storage
 
         if callable(self.storage):
-            self.storage = self.storage()
+            check_storage = self.storage()
+            if not issubclass(check_storage, Storage):
+                raise ImproperlyConfigured("%s.storage must be a subclass of %s.%s ()" % (self.__class__.__name__, Storage.__module__, Storage.__name__))
+            self.storage = check_storage
```
Comment (carltongibson) [inline django/db/models/fields/files.py]: * You need to use `isinstance` here not `issubclass`. 
* I don't think we need the temporary variable. Just assign to `self.storage` as before and then check that.

Comment (miigotu) [reply django/db/models/fields/files.py]: Yeah, it is instantiated, I need to adjust a bit still. It has to be issubclass though.

**Comment by miigotu** [comment]
@carltongibson I will squash and reword the commit after this build passes, then I think it is ready for a final review on success.

**Comment by miigotu** [comment]
This is going to take me a bit more work/thought. These tests fail on master for me also.

**Comment by miigotu** [comment]
@carltongibson I will add docstrings/docs tomorrow-ish. Please let me know if you can think of any other tests I can add or anything else to improve this. Thanks

**Comment by miigotu** [comment]
Let me know if this needs more work, thanks.

**Comment by miigotu** [comment]
Will do

File: django/db/models/fields/files.py [hunk 5]
```diff
@@ -224,6 +225,15 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self._primary_key_set_explicitly = 'primary_key' in kwargs
 
         self.storage = storage or default_storage
+
+        if callable(self.storage):
+            self.storage = self.storage()
```
Comment (Jafnee) [inline django/db/models/fields/files.py]: Would it be possible to have the instance passed to the callable? Similar to `upload_to`

Comment (miigotu) [reply django/db/models/fields/files.py]: An instance has always been possible.
I totally forgot to fix this PR, I'll try and do it tonight to get it merged.

File: tests/file_storage/tests.py (hunk 3 above)
Comment (miigotu) [reply tests/file_storage/tests.py]: All done, tests running now

File: tests/file_storage/tests.py (hunk 2 above)
Comment (miigotu) [reply tests/file_storage/tests.py]: Ive added a "test_callable_base_class_error_raises" test for when the callable is not a sub of Storage

**Comment by miigotu** [comment]
@carltongibson I wasn't sure where to add to the changelog. Does this have to wait until 2.2 now or is it allowed in 2.1.3?
Also, can you resolve/update your review please? I broke github by squashing and force push lol, but I did do all of the requested changes to the PR so far, including some extra tests.
I also removed the changes I made to the wordlist, because I don't think they should be in this patch.

File: tests/file_storage/tests.py [hunk 6]
```diff
@@ -842,6 +847,48 @@ def test_stringio(self):
...
+    def test_callable_base_class_error_raises(self):
+        class NotStorage:
+            pass
+        msg = "FileField.storage must be a subclass/instance of django.core.files.storage.Storage"
+        with self.assertRaisesMessage(ImproperlyConfigured, msg):
+            FileField(storage=NotStorage)
+            FileField(storage=None)
+            FileField(storage=str)
+            FileField(storage=list)
+            FileField(storage=set)
```
Comment (carltongibson) [inline tests/file_storage/tests.py]: Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

Comment (carltongibson) [inline tests/file_storage/tests.py]: Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

File: docs/topics/files.txt [hunk 7]
```diff
@@ -181,3 +181,31 @@ For example, the following code will store uploaded files under
 :doc:`Custom storage systems </howto/custom-file-storage>` work the same way:
 you can pass them in as the ``storage`` argument to a
 :class:`~django.db.models.FileField`.
+
+Callable file storage
```
Comment (carltongibson) [inline docs/topics/files.txt]: Is this quite right? 

It's not that the file storage is callable but that the callable provides the storage... Maybe just "Using a callable" or "...to specify the storage"? 

(Feels a bit nit-picky but...)

File: tests/file_storage/tests.py (hunk 6 above)
Comment (carltongibson) [inline tests/file_storage/tests.py]: Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

File: docs/topics/files.txt (hunk 7 above)
Comment (carltongibson) [inline docs/topics/files.txt]: Feels a bit nit-picky but, is this quite right? 

It's not that the storage is callable, but that the callable provides the storage, so maybe just "Using a callable" or "...to specify the storage" ?

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

**Comment by carltongibson** [comment]
Sorry about the duplicate reviews. GH's UI was having difficulties.

**Comment by carltongibson** [comment]
Hi @miigotu — GH's review panel seems to be failing me this morning... I'll try a comment 😄 

Sorry if this ends up appearing several times. 

----

Hi @miigotu

Thanks for updating here. It's looking good. 👍 

> Does this have to wait until 2.2 now or is it allowed in 2.1.3?

It's a new feature so, it'll have to go in 2.2. Can you update accordingly? 


> `docs/topics/files.txt:185`
> Feels a bit nit-picky but, is this quite right? 
> 
> It's not that the storage is callable, but that the callable provides the storage... Maybe "Using a callable" or "...to specify the storage". ?


> `tests/file_storage/tests.py:866`
> Would this block be better using [`subTest()`](https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.subTest)?

File: docs/topics/files.txt (hunk 7 above)
Comment (miigotu) [reply docs/topics/files.txt]: I can change the verbage there. Github is having some issues right now - https://status.github.com/messages

File: django/db/models/fields/files.py (hunk 5 above)
Comment (miigotu) [reply django/db/models/fields/files.py]: Fun thing here I just realized.

You can now pass an instance of Storage with an added __call__ method:
```python
class MyStorage(Storage):
    def __call__(self):
        if settings.BLAH:
            self.location = '...'
        return self

my_image = models.ImageField(location=MyStorage())
````

File: docs/releases/2.2.txt [hunk 8]
```diff
@@ -148,7 +148,10 @@ Email
 File Storage
 ~~~~~~~~~~~~
 
-* ...
+* Added the ability to pass a callable as the storage parameter of a :class:`~django.db.models.FileField`, or
```
Comment (auvipy) [inline docs/releases/2.2.txt]: move this part to 3.0

**Comment by LuRsT** [comment]
@miigotu Need help with this PR?

**Comment by nikolaik** [comment]
Would have used this feature today if was merged. Any help wanted?

**Comment by miigotu** [comment]
Yes. If you can get it ready I will accept PR to my repo. It will need rebased and updated.

File: docs/releases/2.2.txt (hunk 8 above)
Comment (miigotu) [reply docs/releases/2.2.txt]: @nikolaik It needs to be taken out of 2.2 release also.

**Comment by miigotu** [comment]
@carltongibson I'm not sure exactly what you mean? A failure case or a success case? Is it more to exemplify how someone could have a subclass of Storage that also has a `__call__` method that returns an already instantiated storage attribute?

I guess I just don't understand the goal of that one, we currently handle it the same way any other place a callable is allowed to be passed into the model manager.

**Comment by carltongibson** [comment]
Hi @miigotu. I pushed a fixup removing the callable storage example. I think it's probably clearer without it. 

I'll ask @felixxm to have a look here.

File: docs/ref/models/fields.txt [hunk 9]
```diff
@@ -822,8 +822,9 @@ Has two optional arguments:
 
 .. attribute:: FileField.storage
 
-    A storage object, which handles the storage and retrieval of your
-    files. See :doc:`/topics/files` for details on how to provide this object.
+    A storage object, or a callable which returns a storage object. This
+    handles the storage and retrieval of your files. See :doc:`/topics/files`
+    for details on how to provide this object.
```
Comment (carltongibson) [inline docs/ref/models/fields.txt:827]: Needs a `.. versionchanged:`

File: docs/topics/files.txt [hunk 10]
```diff
@@ -202,3 +202,31 @@ For example, the following code will store uploaded files under
 :doc:`Custom storage systems </howto/custom-file-storage>` work the same way:
 you can pass them in as the ``storage`` argument to a
 :class:`~django.db.models.FileField`.
+
+Using a callable
+----------------
+
```
Comment (carltongibson) [inline docs/topics/files.txt:208]: Could put in a `.. versionchanged:` here?

File: django/db/models/fields/files.py [hunk 11]
```diff
@@ -234,6 +235,15 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
...
+            if not isinstance(self.storage, Storage):
+                raise ImproperlyConfigured(
+                    "%s.storage must be a subclass/instance of %s.%s"
+                    % (self.__class__.__name__, Storage.__module__, Storage.__name__)
```
Comment (felixxm) [inline django/db/models/fields/files.py:244]: `__name__` -> `__qualname__`

File: django/db/models/fields/files.py [hunk 12]
```diff
@@ -234,6 +235,15 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self._primary_key_set_explicitly = 'primary_key' in kwargs
 
         self.storage = storage or default_storage
+
+        if callable(self.storage):
+            self.storage = self.storage()
+            if not isinstance(self.storage, Storage):
+                raise ImproperlyConfigured(
```
Comment (felixxm) [inline django/db/models/fields/files.py:242]: Maybe `TypeError` :thinking:

File: django/db/models/fields/files.py [hunk 13]
```diff
@@ -234,6 +235,15 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
...
+                    "%s.storage must be a subclass/instance of %s.%s"
+                    % (self.__class__.__name__, Storage.__module__, Storage.__name__)
+                )
+
```
Comment (felixxm) [inline django/db/models/fields/files.py:246]: Chop blank line.

File: docs/releases/3.1.txt [hunk 14]
```diff
@@ -248,6 +248,12 @@ File Storage
 
 * ``FileSystemStorage.save()`` method now supports :class:`pathlib.Path`.
 
+* Added the ability to pass a callable as the storage parameter of a
+  :class:`~django.db.models.FileField`, or
+  :class:`~django.db.models.ImageField`.This allows you to modify the used
```
Comment (felixxm) [inline docs/releases/3.1.txt:253]: Maybe:
```
* :class:`~django.db.models.FileField` and class:`~django.db.models.ImageField`
  now accept a callable for ``storage``. This ...
```

File: docs/releases/3.1.txt [hunk 15]
```diff
@@ -248,6 +248,12 @@ File Storage
...
+  storage at runtime, selecting different storages for development and
```
Comment (felixxm) [inline docs/releases/3.1.txt:254]: ` and` -> `, and` (serial comma)

File: docs/topics/files.txt (hunk 10 above)
Comment (felixxm) [reply docs/topics/files.txt:208]: It's a new section, so I would add a `versionadded` at the beginning:
```
Using a callable
----------------

.. versionadded:: 3.1

You can use ...
```

File: docs/topics/files.txt [hunk 16]
```diff
@@ -202,3 +202,31 @@ For example, the following code will store uploaded files under
...
+For example::
+
+    import os
+    from django.db import models
+    from .storages import MyLocalStorage, MyRemoteStorage
+
+
+    def select_storage():
+        if os.getenv('IS_PRODUCTION', False):
```
Comment (felixxm) [inline docs/topics/files.txt:225]: Can we use `settings.DEBUG`?
```python
return MyLocalStorage() if settings.DEBUG else MyRemoteStorage()
```

File: tests/file_storage/models.py [hunk 17]
```diff
@@ -44,6 +52,10 @@ def pathlib_upload_to(self, filename):
         storage=CustomValidNameStorage(location=temp_storage_location),
         upload_to=random_upload_to,
     )
+
+    callable_storage_field = models.FileField(storage=callable_storage, upload_to='callable_storage')
+    callable_storage_field2 = models.FileField(storage=CallableStorage, upload_to='callable_storage2')
+
```
Comment (felixxm) [inline tests/file_storage/models.py:58]: Chop blank lines.

File: tests/file_storage/tests.py [hunk 18]
```diff
@@ -866,6 +872,52 @@ def test_stringio(self):
...
+        self.temp_storage_location = tempfile.mkdtemp(suffix='filefield_callable_storage')
+
+    def tearDown(self):
+        shutil.rmtree(self.temp_storage_location)
+
+    def test_callable_base_class_error_raises(self):
+        class NotStorage:
+            pass
+        msg = "FileField.storage must be a subclass/instance of django.core.files.storage.Storage"
+        for type_ in (NotStorage, str, list, set, tuple):
```
Comment (felixxm) [inline tests/file_storage/tests.py:886]: `type_` -> `invalid_type`

File: tests/file_storage/tests.py [hunk 19]
```diff
@@ -866,6 +872,52 @@ def test_stringio(self):
...
+    def setUp(self):
+        self.temp_storage_location = tempfile.mkdtemp(suffix='filefield_callable_storage')
+
+    def tearDown(self):
+        shutil.rmtree(self.temp_storage_location)
+
+    def test_callable_base_class_error_raises(self):
+        class NotStorage:
+            pass
+        msg = "FileField.storage must be a subclass/instance of django.core.files.storage.Storage"
```
Comment (felixxm) [inline tests/file_storage/tests.py:885]: Use single quotes.

File: tests/file_storage/tests.py [hunk 20]
```diff
@@ -866,6 +872,52 @@ def test_stringio(self):
...
+        for type_ in (NotStorage, str, list, set, tuple):
+            with self.subTest(type_=type_):
+                with self.assertRaisesMessage(ImproperlyConfigured, msg):
+                    FileField(storage=type_)
+
+        # These should not raise, and should use default_storage
+        for type_ in (None, False):
+            with self.subTest(type_=type_):
+                obj = FileField(storage=type_)
+                self.assertEqual(obj.storage, default_storage)
```
Comment (felixxm) [inline tests/file_storage/tests.py:895]: These assertions are not related with this patch. Personally I don't see much value in adding them.

File: tests/file_storage/models.py [hunk 21]
```diff
@@ -44,6 +52,10 @@ def pathlib_upload_to(self, filename):
         storage=CustomValidNameStorage(location=temp_storage_location),
         upload_to=random_upload_to,
     )
+
+    callable_storage_field = models.FileField(storage=callable_storage, upload_to='callable_storage')
```
Comment (felixxm) [inline tests/file_storage/models.py:56]: Maybe `callable_storage_field` -> `storage_callable`

File: tests/file_storage/models.py [hunk 22]
```diff
@@ -44,6 +52,10 @@ def pathlib_upload_to(self, filename):
...
+    callable_storage_field2 = models.FileField(storage=CallableStorage, upload_to='callable_storage2')
```
Comment (felixxm) [inline tests/file_storage/models.py:57]: Maybe `callable_storage_field2` -> `storage_callable_class`

File: docs/topics/files.txt [hunk 23]
```diff
@@ -202,3 +202,31 @@ For example, the following code will store uploaded files under
 :doc:`Custom storage systems </howto/custom-file-storage>` work the same way:
 you can pass them in as the ``storage`` argument to a
 :class:`~django.db.models.FileField`.
+
+Using a callable
+----------------
+
+You can use a callable as the storage parameter for
```
Comment (felixxm) [inline docs/topics/files.txt:209]: `storage` -> `` :attr:`~FileField.storage` ``

File: docs/topics/files.txt [hunk 24]
```diff
@@ -202,3 +202,31 @@ For example, the following code will store uploaded files under
...
+:class:`~django.db.models.FileField` or :class:`~django.db.models.ImageField`.
+This allows you to modify the used storage at runtime, selecting different
+storages for development and production environments, for example.
+
+Your callable will be evaluated when your models classes are loaded, and must
```
Comment (felixxm) [inline docs/topics/files.txt:214]: I would chop the second `your`.

File: tests/file_storage/tests.py (hunk 20 above)
Comment (miigotu) [reply tests/file_storage/tests.py:895]: It does cover a previously untested scenario though, afaik. This assures that this and any future patch does not break this behavior retroactively.

**Comment by miigotu** [comment]
@carltongibson Ahh, the callable storage example I forgot about that. So that is what you wanted a test case around, lol. I think it's probably clearer without it also, but it is one of the primary use cases where storage would adapt automagically ;)

I'll try and get some of these documentation changes in over the next few days f nobody else gets to it. I am shooting in the dark when doing this style of documentation since its my first time.

File: tests/file_storage/tests.py (hunk 20 above)
Comment (felixxm) [reply tests/file_storage/tests.py:895]: `False` is not an expected value, it works as a side effect of the current implementation. We shouldn't add tests for passing `False` because it can create an impression that this is expected and supported.

About `None`, I would rather add test for
```python
self.assertEqual(FileField().storage, default_storage)
```
Nevertheless it's not related with this new feature and should be moved to a separate commit.
